In [2]:
install.packages("e1071")
install.packages("caret")
library(e1071)
library(caret)

package 'e1071' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\marc_\AppData\Local\Temp\RtmpW6l7fo\downloaded_packages
package 'caret' successfully unpacked and MD5 sums checked

The downloaded binary packages are in
	C:\Users\marc_\AppData\Local\Temp\RtmpW6l7fo\downloaded_packages


Warning message:
"package 'e1071' was built under R version 3.6.3"Warning message:
"package 'caret' was built under R version 3.6.3"Loading required package: lattice
Warning message:
"package 'lattice' was built under R version 3.6.3"Loading required package: ggplot2
Warning message:
"package 'ggplot2' was built under R version 3.6.3"

In [30]:
Titanic<-as.data.frame(Titanic)
Titanic

Class Sex    Age   Survived Freq
1  1st   Male   Child No         0 
2  2nd   Male   Child No         0 
3  3rd   Male   Child No        35 
4  Crew  Male   Child No         0 
5  1st   Female Child No         0 
6  2nd   Female Child No         0 
7  3rd   Female Child No        17 
8  Crew  Female Child No         0 
9  1st   Male   Adult No       118 
10 2nd   Male   Adult No       154 
11 3rd   Male   Adult No       387 
12 Crew  Male   Adult No       670 
13 1st   Female Adult No         4 
14 2nd   Female Adult No        13 
15 3rd   Female Adult No        89 
16 Crew  Female Adult No         3 
17 1st   Male   Child Yes        5 
18 2nd   Male   Child Yes       11 
19 3rd   Male   Child Yes       13 
20 Crew  Male   Child Yes        0 
21 1st   Female Child Yes        1 
22 2nd   Female Child Yes       13 
23 3rd   Female Child Yes       14 
24 Crew  Female Child Yes        0 
25 1st   Male   Adult Yes       57 
26 2nd   Male   Adult Yes       14 
27 3rd   Male   Adult Yes       75 
28 Crew  Male   Adult Yes      192 
29 1st   Female Adult Yes      140 
30 2nd   Female Adult Yes       80 
31 3rd   Female Adult Yes       76 
32 Crew  Female Adult Yes       20

In [31]:
summary(Titanic)

  Class       Sex        Age     Survived      Freq       
 1st :8   Male  :16   Child:16   No :16   Min.   :  0.00  
 2nd :8   Female:16   Adult:16   Yes:16   1st Qu.:  0.75  
 3rd :8                                   Median : 13.50  
 Crew:8                                   Mean   : 68.78  
                                          3rd Qu.: 77.00  
                                          Max.   :670.00  

Podemos observar en los datos que tenemos un total de 32 datos, cada dato tiene asignado un sexo, edad, si sobrevivió o no y la frecuencia. Este último parámetro indica el número de personas con esos parámetros que había en el Titanic. De esta manera, hay que ajustar el dataset para tener la cantidad de datos dependiendo de las frecuencias, asi poder realizar el algoritmo de la mejor manera y tener los resultados más fiables respecto a la realidad.

### Creación del DATASET 

In [32]:
#Ahora repetiremos cada combinación igual a la frecuencia de cada fila
Frecuencia=rep.int(seq_len(nrow(Titanic)), Titanic$Freq) 
#Creamos el nuevo dataset teniendo encuenta la frecuencia
Titanic_dataset=Titanic[Frecuencia,]
#Commo ya hemos tenido en cuenta la frecuencia la podemos eliminar del dataset
Titanic_dataset$Freq=NULL
Titanic_dataset
summary(Titanic_dataset)

Class Sex    Age   Survived
3     3rd   Male   Child No      
3.1   3rd   Male   Child No      
3.2   3rd   Male   Child No      
3.3   3rd   Male   Child No      
3.4   3rd   Male   Child No      
3.5   3rd   Male   Child No      
3.6   3rd   Male   Child No      
3.7   3rd   Male   Child No      
3.8   3rd   Male   Child No      
3.9   3rd   Male   Child No      
3.10  3rd   Male   Child No      
3.11  3rd   Male   Child No      
3.12  3rd   Male   Child No      
3.13  3rd   Male   Child No      
3.14  3rd   Male   Child No      
3.15  3rd   Male   Child No      
3.16  3rd   Male   Child No      
3.17  3rd   Male   Child No      
3.18  3rd   Male   Child No      
3.19  3rd   Male   Child No      
3.20  3rd   Male   Child No      
3.21  3rd   Male   Child No      
3.22  3rd   Male   Child No      
3.23  3rd   Male   Child No      
3.24  3rd   Male   Child No      
3.25  3rd   Male   Child No      
3.26  3rd   Male   Child No      
3.27  3rd   Male   Child No      
3.28  3rd   Male   Child No      
3.29  3rd   Male   Child No      
...   ...   ...    ...   ...     
31.66 3rd   Female Adult Yes     
31.67 3rd   Female Adult Yes     
31.68 3rd   Female Adult Yes     
31.69 3rd   Female Adult Yes     
31.70 3rd   Female Adult Yes     
31.71 3rd   Female Adult Yes     
31.72 3rd   Female Adult Yes     
31.73 3rd   Female Adult Yes     
31.74 3rd   Female Adult Yes     
31.75 3rd   Female Adult Yes     
32    Crew  Female Adult Yes     
32.1  Crew  Female Adult Yes     
32.2  Crew  Female Adult Yes     
32.3  Crew  Female Adult Yes     
32.4  Crew  Female Adult Yes     
32.5  Crew  Female Adult Yes     
32.6  Crew  Female Adult Yes     
32.7  Crew  Female Adult Yes     
32.8  Crew  Female Adult Yes     
32.9  Crew  Female Adult Yes     
32.10 Crew  Female Adult Yes     
32.11 Crew  Female Adult Yes     
32.12 Crew  Female Adult Yes     
32.13 Crew  Female Adult Yes     
32.14 Crew  Female Adult Yes     
32.15 Crew  Female Adult Yes     
32.16 Crew  Female Adult Yes     
32.17 Crew  Female Adult Yes     
32.18 Crew  Female Adult Yes     
32.19 Crew  Female Adult Yes

  Class         Sex          Age       Survived  
 1st :325   Male  :1731   Child: 109   No :1490  
 2nd :285   Female: 470   Adult:2092   Yes: 711  
 3rd :706                                        
 Crew:885                                        

### División del DATASET - Train y Test

In [33]:
set.seed(2021)
tamano.total=nrow(Titanic_dataset)
tamano.train=round(tamano.total*0.75)

Titanic_dataset.indices = sample(1:tamano.total , size=tamano.train)
Titanic_dataset.train = Titanic_dataset[Titanic_dataset.indices,]
Titanic_dataset.test = Titanic_dataset[-Titanic_dataset.indices,]
print("Este es el TRAIN")
summary(Titanic_dataset.train)
print("Este es el TEST")
summary(Titanic_dataset.test)

[1] "Este es el TRAIN"


  Class         Sex          Age       Survived  
 1st :252   Male  :1291   Child:  84   No :1105  
 2nd :209   Female: 360   Adult:1567   Yes: 546  
 3rd :520                                        
 Crew:670                                        

[1] "Este es el TEST"


  Class         Sex         Age      Survived 
 1st : 73   Male  :440   Child: 25   No :385  
 2nd : 76   Female:110   Adult:525   Yes:165  
 3rd :186                                     
 Crew:215                                     

###  Clasificador Naive Bayes para la variable objetivo "Survived"

In [34]:
modelo = naiveBayes(Survived ~ ., data=Titanic_dataset.train) 
modelo


Naive Bayes Classifier for Discrete Predictors

Call:
naiveBayes.default(x = X, y = Y, laplace = laplace)

A-priori probabilities:
Y
       No       Yes 
0.6692913 0.3307087 

Conditional probabilities:
     Class
Y            1st        2nd        3rd       Crew
  No  0.08144796 0.10678733 0.34841629 0.46334842
  Yes 0.29670330 0.16666667 0.24725275 0.28937729

     Sex
Y           Male     Female
  No  0.91312217 0.08687783
  Yes 0.51648352 0.48351648

     Age
Y          Child      Adult
  No  0.03348416 0.96651584
  Yes 0.08608059 0.91391941


### Predicción

In [35]:
results = predict(object = modelo, newdata=Titanic_dataset.test, type = "class")
results

[1] No  No  No  No  No  No  No  No  No  No  Yes Yes Yes Yes Yes No  No  No 
 [19] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
 [37] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
 [55] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
 [73] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
 [91] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[109] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[127] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[145] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[163] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[181] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[199] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[217] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[235] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[253] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[271] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[289] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[307] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[325] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[343] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[361] Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
[379] Yes Yes Yes Yes Yes Yes Yes Yes Yes No  No  No  Yes Yes Yes Yes Yes No 
[397] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[415] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[433] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[451] No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No  No 
[469] No  No  No  No  No  No  No  Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
[487] Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
[505] Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
[523] Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
[541] Yes Yes Yes Yes Yes Yes Yes Yes Yes Yes
Levels: No Yes

### Visualización de resultados

In [36]:
tab = table(Titanic_dataset.test$Survived,results,dnn = c('REAL','PREDICCIÓN'))
confusionMatrix(tab) 

Confusion Matrix and Statistics

     PREDICCIÓN
REAL   No Yes
  No  355  30
  Yes  83  82
                                          
               Accuracy : 0.7945          
                 95% CI : (0.7583, 0.8276)
    No Information Rate : 0.7964          
    P-Value [Acc > NIR] : 0.5671          
                                          
                  Kappa : 0.4614          
                                          
 Mcnemar's Test P-Value : 9.994e-07       
                                          
            Sensitivity : 0.8105          
            Specificity : 0.7321          
         Pos Pred Value : 0.9221          
         Neg Pred Value : 0.4970          
             Prevalence : 0.7964          
         Detection Rate : 0.6455          
   Detection Prevalence : 0.7000          
      Balanced Accuracy : 0.7713          
                                          
       'Positive' Class : No              
                                          

In [55]:
tabla1 <- table(Titanic_dataset$Class, Titanic_dataset$Survived, dnn = c("Clase", "Survived"))
tabla1
tabla2 <- table(Titanic_dataset$Sex, Titanic_dataset$Survived, dnn = c("Sexo", "Survived"))
tabla2
tabla3 <- table(Titanic_dataset$Age, Titanic_dataset$Survived, dnn = c("Años", "Survived"))
tabla3
tabla4 <- table(Titanic_dataset$Age, Titanic_dataset$Class, dnn = c("Años", "Clase"))
tabla4

      Survived
Clase   No Yes
  1st  122 203
  2nd  167 118
  3rd  528 178
  Crew 673 212

        Survived
Sexo       No  Yes
  Male   1364  367
  Female  126  344

       Survived
Años      No  Yes
  Child   52   57
  Adult 1438  654

       Clase
Años    1st 2nd 3rd Crew
  Child   6  24  79    0
  Adult 319 261 627  885

# Análisis y Conclusiones

A primera vista nos indica que el modelo tiene una "Accuracy" de 0,7954 que sería un 79,45% no parece del todo malo.
Si nos fijamos en los casos de "No Survived" apreciamos que de los 385 casos totales clasifica bien 355 y 30 mal. Esto supone una "Accuracy" del 92,2% (Muy buena). Por otra parte, si nos centramos en la predicción de los "Yes Survived" la "Accuracy" desciende mucho. Apreciamos que hay 165 casos que "Yes Survived" de los cuales, clasifica bien 82 y clasifica mal 83. Esto supone una "Accuracy" de 49,7% (Bastante mala).

Al observar las tablas hechas y el clasificador Naive-Bayes podemos apreciar como si eres niño, mujer o viajas en primera clase tienes más probabilidades de sobrevivir.

Naive Bayes es un algoritmo paramétrico, esto implica que, aunque hagamos muchas repeticiones, los resultados seguiran siendo los mismos. La única manera de mejorar los resultados o que sean diferentes, sería tener más parámetros a analizar o más cantidad de datos. También hay que tener en cuenta que los parámetros que analizamos para hacer el modelo son independientes entre si. A primera vista debería ser así, pero diría que el viajar a en primera clase o no parámetro muy decisivo en si sobrevives o no depende bastante de la edad. Ya que, proporcionalmente si eres adulto es más probable que viajes en primera clase.

95% CI : (0.7583, 0.8276) Intervalo de confianza Que nos muestra que la "Accuracy" de este modelo va del 75,8% al 82,8%

No Information Rate : 0.7964 esta es la precisión que se puede lograr al predecir siempre la etiqueta de la clase mayoritaria. En este caso, si se le pide que prediga si una persona sobrevive o no, al elegir siempre "No Survived", podemos lograr casi un 80% de precisión en el conjunto de prueba. Nuestra "Accuracy" no se ve tan bien ahora.

P-Value [Acc > NIR] : 0.5671  este p-value muestra que el modelo no ofrece un rendimiento significativo  sobre la tasa de ausencia de información.

Kappa : 0.4614 el estadístico de Kappa muestra qué tan bien nuestras predicciones de clasificadores coincidieron con las etiquetas de clase reales mientras se controla la precisión de un clasificador aleatorio. El valor de 0,46 representa una moderada coincidencia. Contra más cerca de 1 este este valor mejor.

Mcnemar's Test P-Value : 9.994e-07 Nos muestra la probabilidad de estar equivocados si concluimos que los parámetros que utilizamos para hacer el modelo para calcular la supervivencia están asociados al parámetro "Survived"

Sensitivity : 0.8105 se calcula como el número de predicciones "No survived" correctas dividido por el número total de "No Survived". Contra más cerca de 1 se encuentre mejor más sensibilidad tendrá nuestro modelo.

Specificity : 0.7321 se calcula como el número de predicciones "Yes Survived" correctas dividido por el número total de "Yes Survived". La mejor especifidad es un valor de 1. 

Pos Pred Value : 0.9221 se calcula como el número de predicciones positivas correctas dividido por el número total de predicciones positivas. También se le llama valor predictivo positivo. En este caso el PPV está calculado respecto al "No Survivied"

Neg Pred Value : 0.4970 se calcula como el número de predicciones positivas incorrectas dividido por el número total de negativas. En este caso el NPV está calculado respecto al "Yes Survived"

Prevalence : 0.7964 muestra la frecuencia en que aparecen los "No Survived" en este modelo.  

Detection Rate : 0.6455 muestra el número de predicciones de "No Survived" correctas hechas como una proporción de todas las predicciones hechas

Detection Prevalence : 0.7000 muestra el número de predicciones de clase "No Survived" hechas como proporción de todas las predicciones

Balanced Accuracy : 0.7713  esencialmente toma el promedio de las tasas verdaderas positivas y verdaderas negativas

